Add sp_helpdbfixedrole

wboettge opened the pull request:
### Description

Adds the sp_helpdbfixedrole stored procedure. The procedure takes a name of a fixed database role, and returns information about that role. If no role is specified return a list and description of all fixed database roles.

Currently, db_owner is the only fixed database-level role from SQL-Server in Babelfishdb_owner. Like sp_helpsrvrolemember in BABEL-3462, an empty result set will be returned if the procedure is called with any of the other 8 unsupported fixed db roles.

If called with any input besides one of SQL-Server's fixed db roles, an error is raised.
 
### Issues Resolved

Task: BABEL-3086

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Walt Boettge <wboettge@amazon.com>

wboettge committed:
Add sp_helpdbfixedrole

Adds the sp_helpdbfixedrole stored procedure. The procedure takes a name
of a fixed database role, and returns information about that role. If no
role is specified return a list and description of all fixed database
roles.

Currently, db_owner is the only fixed database-level role from
SQL-Server in Babelfishdb_owner. Like sp_helpsrvrolemember in
BABEL-3462, an empty result set will be returned if the procedure is
called with any of the other 8 unsupported fixed db roles.

If called with any input besides one of SQL-Server's fixed db roles, an
error is raised.

Task: BABEL-3086

Signed-off-by: Walt Boettge <wboettge@amazon.com>

wboettge commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2516,6 +2516,36 @@ $$
+		FROM (VALUES (CAST(NULL AS sys.SYSNAME), CAST(NULL AS sys.nvarchar(70))))
+			AS t (DbFixedRole, Description) 
+		WHERE 1=0	
+	END
```
Is there a better way to return an empty result set?

xhfanhe replied:
That's a good question and I don't know the answer... for now I think this is acceptable

xhfanhe commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2516,6 +2516,36 @@ $$
+$$
+LANGUAGE 'pltsql';
+GRANT EXECUTE ON PROCEDURE sys.sp_helpdbfixedrole TO PUBLIC;
+
```
Please add this part to the upgrade script as well.

xhfanhe commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2516,6 +2516,36 @@ $$
+		WHERE 1=0	
+	END
+	ELSE
+		RAISERROR('%s is not a known fixed role.', 16, 1, @rolename);
```
Please add single quotation marks around the rolename

wboettge replied:
Will do

xhfanhe commented on test/JDBC/input/storedProcedures/Test-sp_helpdbfixedrole-vu-prepare.sql:
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE test_sp_helpdbfixedrole_tbl (DbFixedRole sys.SYSNAME, 
```
I think we don't need a table to store the procedure output because this time we don't have a flexible value column like SID in sp_helpsrvrolemember.

xhfanhe commented on test/JDBC/input/storedProcedures/Test-sp_helpdbfixedrole-vu-verify.sql:
```diff
@@ -0,0 +1,55 @@
+GO
+
+EXEC sp_helpdbfixedrole 'error'
+GO
```
Since this is a new test file, you'll need to add it to upgrade test schedule. In `upgrade` folder, for each version (13_4 to latest), please add Test-sp_helpdbfixedrole to the schedule.

Another thing is that, we need another upgrade test Test-sp_helpdbfixedrole-dep, you can refer to `input/storedProcedures/Test-sp_helpsrvrolemember-dep`. The dep test is to test object dependency for releases after this cycle. 
In this dep test, we need to prepare one procedure, one function and one view that depend on sp_helpdbfixedrole.
For now the dep test doesn't need to be added into any schedule. At the start of next cycle we'll add it to latest/schedule.

wboettge committed:
Add Dependency Tests and other fixes

- Add Test-sp_helpdbfixedrole-dep-vu to test dependencies on
  helpdbfixedrole
- Remove use of local variable from implementation that was causing
  issues in tests. See BABEL-3564.
- Add procedure to upgrade script
- Remove unnecessary steps from Test-sp_helpdbfixedrole-vu. Prepare and
  cleanup steps are now empty
- Add Test-sp_helpdbfixedrole-vu to test schedules

Signed-off-by: Walt Boettge <wboettge@amazon.com>

wboettge committed:
Change implementation in upgrade script to match

sp_helpdbfixedrole in 2.2--2.3 upgrade script implementation now matches
code in babelfishpg_tsql.

Signed-off-by: Walt Boettge <wboettge@amazon.com>

xhfanhe commented on test/JDBC/expected/Test-sp_helpdbfixedrole-vu-verify.out:
```diff
@@ -0,0 +1,93 @@
+EXEC sp_helpdbfixedrole
```
Can we prepare a table in prepare script, and then insert to it in verify script? 
We don't have to do the INSERT EXEC within a procedure or function, instead we directly call it (that should not lead to failure right?)
This way at least we're not testing on a purely empty database.

wboettge replied:
For sure, and even if we were doing this within a function this wouldn't lead to any issues since there's no longer any local variables in the implementation.

wboettge committed:
Update tests to include inserting into table

Signed-off-by: Walt Boettge <wboettge@amazon.com>

wboettge committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3086
